packages/cli/src/ui/utils/textOutput.test.ts

```diff
@@ -0,0 +1,77 @@
+/**
+ * @license
+ * Copyright 2025 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { vi } from 'vitest';
+import { textOutput } from './textOutput.js';
+
+describe('textOutput', () => {
+  let stdoutSpy: vi.SpyInstance;
+
+  beforeEach(() => {
+    stdoutSpy = vi
+      .spyOn(process.stdout, 'write')
+      .mockImplementation(() => true);
+    textOutput._resetForTesting();
+  });
+
+  afterEach(() => {
+    stdoutSpy.mockRestore();
+  });
+
+  it('write() should call process.stdout.write', () => {
+    textOutput.write('hello');
+    expect(stdoutSpy).toHaveBeenCalledWith('hello');
+  });
+
+  it('write() should not call process.stdout.write for empty strings', () => {
+    textOutput.write('');
+    expect(stdoutSpy).not.toHaveBeenCalled();
+  });
+
+  it('writeOnNewLine() should not add a newline if the last char was a newline', () => {
+    // Default state starts with a newline
+    textOutput.writeOnNewLine('hello');
+    expect(stdoutSpy).toHaveBeenCalledWith('hello');
```

Comment:
for these tests the easiest way is to just check that the final output to the fae stdout matches an expected snapshot. that is simpler for people to verify the output in the tests. See examples in the codebase of that.